File: Grad_school/FSM/C++_version/FSM.cpp
/*****************************************************************
* Program: DFA Simulator
* Author: Sajjad
* 
* Purpose:
* This program simulates a Deterministic Finite Automaton (DFA).
* It reads a DFA definition from a file (DFA.txt), then evaluates
* input strings provided by the user. It outputs a computation 
* trace for each string and indicates whether the string is accepted
* or rejected by the DFA. The program handles invalid input 
* characters and can be terminated by entering "Quit".
* 
* Input:
* The DFA is loaded from a file in the following format:
* - Line 1: Number of states.
* - Line 2: Space-separated list of accepting states.
* - Line 3: Space-separated alphabet characters.
* - Lines 4+: Transition table for each state and alphabet character.
* 
* Output:
* The program prints the DFA computation trace for each string 
* and whether the string is "ACCEPTED" or "REJECTED".
* 
*****************************************************************/

#include <iostream>                                                                             // cin, cout standard input/output
#include <fstream>                                                                              // file input/output operations
#include <vector>                                                                               // dynamic arrays better than static arrays in this case
#include <set>                                                                                  // set data structure for accepting states
#include <string>                                                                               // string data type for input strings

using namespace std;                                                                            // standard namespace for cin, cout, etc.just C++ thingy

struct DFA {                                                                                    // struct for DFA 
    int num_states;                                                                             // number of states
    set<int> accepting_states;                                                                  // set of accepting states
    vector<char> alphabet;                                                                      // alphabet symbols
    vector<vector<int>> transitions;                                                            // transition table
};

                                                                                                // Function to load the DFA from a file
DFA load_dfa(const string& filename) {                                                          // argument is filename of the file to be loaded
    ifstream file(filename);                                                                    // file object to read the file
    DFA dfa;                                                                                    // DFA object to store the DFA
    if (!file) {                                                                                // error check in case file is not found
        cerr << "Error loading DFA file!" << endl;                                              // print error message and endline
        exit(1);                                                                                // exit the program with error code 1
    }

                                                                                                // First line: number of states
    file >> dfa.num_states;                                                                     // read the number of states from the file

                                                                                                // Second line: accepting states
    int state;                                                                                  // state variable to store the accepting states
    while (file.peek() != '\n') {                                                               // peek() function returns the next character in the input sequence without extracting it
        file >> state;                                                                          // read the state from the file
        dfa.accepting_states.insert(state);                                                     // insert the state into the set of accepting states
    }

                                                                                                // Third line: alphabet symbols
    char symbol;                                                                                // symbol variable to store the alphabet symbols
    while (file >> symbol && symbol != '\n') {                                                  // read the symbol from the file and check if it is not a newline character
        dfa.alphabet.push_back(symbol);                                                         // add the symbol to the alphabet
    }

                                                                                                // Fourth and subsequent lines: transition table
    dfa.transitions.resize(dfa.num_states, vector<int>(dfa.alphabet.size()));                   // resize the transition table to the number of states and alphabet size
    for (int i = 0; i < dfa.num_states; ++i) {                                                  // loop through the states to read the transitions
        for (int j = 0; j < dfa.alphabet.size(); ++j) {                                         // loop through the alphabet to read the transitions
            file >> dfa.transitions[i][j];                                                      // read the transition from the file
        }
    }

    return dfa;                                                                                 // return the DFA object
}

                                                                                                // Function to evaluate the string against the DFA
string evaluate_string(const DFA& dfa, const string& input_string) {                            // arguments are DFA object and input string
    int current_state = 0;                                                                      // current state is initialized to 0 (start state)
    string remaining_input = input_string;                                                      // remaining input is the input string

    cout << ">>>Computation…" << endl;                                                          // print computation... and endline
    for (size_t i = 0; i < input_string.size(); ++i) {                                          // loop through the input string characters
        char current_char = input_string[i];                                                    // current character is the character at index i
        
                                                                                                // Check if the character is part of the alphabet
        auto it = find(dfa.alphabet.begin(), dfa.alphabet.end(), current_char);                 // find the character in the alphabet
        if (it == dfa.alphabet.end()) {                                                         // if the character is not found
            cout << current_state << "," << remaining_input << " -> INVALID INPUT" << endl;     // print the current state, remaining input, and INVALID INPUT
            return "REJECTED";                                                                  // return REJECTED as the input is invalid
        }

        int char_index = distance(dfa.alphabet.begin(), it);                                    // get the index of the character in the alphabet
        int next_state = dfa.transitions[current_state][char_index];                            // get the next state from the transition table
        cout << current_state << "," << remaining_input << " -> " << next_state << "," << remaining_input.substr(1) << endl;            // print the current state, remaining input, next state, and remaining input without the first character

                                                                                                // Update current state and remaining input
        current_state = next_state;                                                             // update the current state to the next state (transition)
        remaining_input = remaining_input.substr(1);                                            // update the remaining input by removing the first character of the input string
    }

                                                                                                // Check if final state is an accepting state
    if (dfa.accepting_states.count(current_state)) {                                            // check if the final state is an accepting state
        return "ACCEPTED";                                                                      // return ACCEPTED if the final state is an accepting state
    } else {                                                                                    // if the final state is not an accepting state
        return "REJECTED";                                                                      // return REJECTED
    }
}

int main() {
                                                                                                // Load the DFA from file
    cout << ">>>Loading DFA.txt…" << endl;                                                      // print Loading DFA.txt and endline
    DFA dfa = load_dfa("DFA.txt");                                                              // load the DFA from the file DFA.txt

    while (true) {                                                                              // infinite loop to run program until user enters Quit
                                                                                                // Prompt user for input string
        cout << ">>>Please enter a string to evaluate: ";                                       // print Please enter a string to evaluate:
        string input_string;                                                                    // input string variable
        getline(cin, input_string);                                                             // get the input string from the user c++ equivalent of scanf in this case

        if (input_string == "Quit") {                                                           // check if the input string is Quit
            cout << ">>>Goodbye!" << endl;                                                      // print Goodbye! and endline
            break;                                                                              // break the loop to exit the program
        }

                                                                                                // Evaluate the input string
        string result = evaluate_string(dfa, input_string);                                     // evaluate the input string against the DFA
        cout << result << endl;                                                                 // print the result (ACCEPTED or REJECTED) and endline
    }

    return 0;                                                                                   // return 0 to indicate successful program execution just like C
}

File: Grad_school/FSM/FSM.py
#****************************************************************
# DFA Simulator
# Author: Sajjad Alsaffar
#
# Purpose: This program simulates a Deterministic Finite Automaton (DFA).
# It reads a DFA definition from a file (DFA.txt) and evaluates strings
# provided by the user. The output is a computation trace and an
# indication of whether the string is accepted or rejected by the DFA.
#****************************************************************

def load_dfa(filename):                                                                     # function to load the DFA from a text file
    """Loads the DFA from a text file."""
    with open(filename, 'r') as file:                                                       # while the file is open as read only do the below
        # First line: number of states
        num_states = int(file.readline().strip())                                           # read the first line and strip the white spaces

        # Second line: accepting states
        accepting_states = set(map(int, file.readline().strip().split()))                   # read the second line and strip the white spaces

        # Third line: alphabet symbols
        alphabet = file.readline().strip().split()                                          # read the third line and strip the white spaces

        # Fourth and subsequent lines: transition table
        transitions = []                                                                    # create an empty list to store the transitions
        for _ in range(num_states):                                                         # loop through the number of states
            transitions.append(list(map(int, file.readline().strip().split())))             # append the transitions to the list and strip the white spaces

    return num_states, accepting_states, alphabet, transitions                              # return the number of states, accepting states, alphabet and transitions

def evaluate_string(dfa, input_string):                                                     # function to evaluate the string takes the DFA and the input string
    """Evaluates whether the input string is accepted or rejected by the DFA."""
    num_states, accepting_states, alphabet, transitions = dfa                               # unpack the DFA
    current_state = 0                                                                       # Start at the initial state

    computation_trace = [(current_state, input_string)]                                     # Store the computation trace in a list

    for char in input_string:                                                               # Loop through the input string
        if char not in alphabet:                                                            # Check if the character is in the alphabet
            print(f"{current_state},{input_string} -> INVALID INPUT")                       # If not the input is not alphabet like 234 or b233, return "REJECTED"
            return "REJECTED"                                                               # Invalid input so return "REJECTED"
        else:                                                                               # If the character is in the alphabet
            char_index = alphabet.index(char)                                               # Get the index of the character in the alphabet
            current_state = transitions[current_state][char_index]                          # Update the current state based on the transition table
            input_string = input_string[1:]                                                 # Remove the first character from the input string by slicing
            computation_trace.append((current_state, input_string))                         # Append the current state and remaining input to the computation trace

    # Output computation trace
    print(">>>Computation…")                                                                # Print the computation....
    for i, (state, remaining_input) in enumerate(computation_trace):                        # Loop through the computation trace by index and the state and remaining input 
        if i < len(computation_trace) - 1:                                                  # if the index is less than the length of the computation trace - 1
            print(f"{state},{remaining_input} -> {computation_trace[i+1][0]},{computation_trace[i+1][1] if computation_trace[i+1][1] else '{e}'}")      # print the state, remaining input and the next state and remaining input
    
    # Check if the final state is an accepting state
    if current_state in accepting_states:                                                   # If the current state is in the accepting states return "ACCEPTED"
        return "ACCEPTED"                                                                   # If the current state is in the accepting states return "ACCEPTED"
    else:
        return "REJECTED"                                                                   # If the current state is not in the accepting states return "REJECTED"

def main():                                                                                 # Main function to run the program
    # Load the DFA from file
    print(">>>Loading DFA.txt...")                                                          # Print the loading message
    dfa = load_dfa('DFA.txt')                                                               # Load the DFA from the file

    while True:                                                                             # Loop until the user enters "quit"
        # Prompt user for input string
        input_string = input(">>>Please enter a string to evaluate: ").strip()              # Prompt the user to enter a string to evaluate and strip the white spaces
        if input_string.lower() == "quit":                                                  # If the user enters "quit" break the loop
            print(">>>Goodbye!")                                                            # Print goodbye message
            break                                                                           # Exit the loop

        # Evaluate the input string
        result = evaluate_string(dfa, input_string)                                         # Evaluate the input string
        print(result)                                                                       # Print the result

if __name__ == "__main__":                                                                  # if the script is run directly
    main()                                                                                  # Run the main function